impeller/renderer/backend/metal/texture_mtl.mm

```diff
@@ -4,120 +4,295 @@
   }
 
-  is_wrapped_ = wrapped;
-  is_valid_ = true;
-}
+  // |Texture|
+  void SetLabel(std::string_view label) override {
+    [texture_ setLabel:@(label.data())];
+  }
+  // |Texture|
+  bool OnSetContents(const uint8_t* contents,
+                     size_t length,
+                     size_t slice) override {
+    if (!IsValid() || !contents || is_wrapped_) {
+      return false;
+    }
 
-std::shared_ptr<TextureMTL> TextureMTL::Wrapper(
-    TextureDescriptor desc,
-    id<MTLTexture> texture,
-    std::function<void()> deletion_proc) {
-  if (deletion_proc) {
-    return std::shared_ptr<TextureMTL>(
-        new TextureMTL(desc, texture, true),
-        [deletion_proc = std::move(deletion_proc)](TextureMTL* t) {
-          deletion_proc();
-          delete t;
-        });
+    const auto& desc = GetTextureDescriptor();
+
+    // Out of bounds access.
+    if (length != desc.GetByteSizeOfBaseMipLevel()) {
+      return false;
+    }
+
+    const auto region =
+        MTLRegionMake2D(0u, 0u, desc.size.width, desc.size.height);
+    [texture_ replaceRegion:region                            //
+                mipmapLevel:0u                                //
+                      slice:slice                             //
+                  withBytes:contents                          //
+                bytesPerRow:desc.GetBytesPerRow()             //
+              bytesPerImage:desc.GetByteSizeOfBaseMipLevel()  //
+    ];
+
+    return true;
   }
-  return std::shared_ptr<TextureMTL>(new TextureMTL(desc, texture, true));
-}
 
-TextureMTL::~TextureMTL() = default;
+  // |Texture|
+  bool OnSetContents(std::shared_ptr<const fml::Mapping> mapping,
+                     size_t slice) override {
+    // Metal has no threading restrictions. So we can pass this data along to
+    // the client rendering API immediately.
+    return OnSetContents(mapping->GetMapping(), mapping->GetSize(), slice);
+  }
 
-void TextureMTL::SetLabel(std::string_view label) {
-  [texture_ setLabel:@(label.data())];
-}
+  // |Texture|
+  bool IsValid() const override { return is_valid_; }
 
-// |Texture|
-bool TextureMTL::OnSetContents(std::shared_ptr<const fml::Mapping> mapping,
-                               size_t slice) {
-  // Metal has no threading restrictions. So we can pass this data along to the
-  // client rendering API immediately.
-  return OnSetContents(mapping->GetMapping(), mapping->GetSize(), slice);
-}
+  // |Texture|
+  ISize GetSize() const override {
+    return {static_cast<ISize::Type>(texture_.width),
+            static_cast<ISize::Type>(texture_.height)};
+  }
+
+  BetterNameTextureMTL(const BetterNameTextureMTL&) = delete;
+
+  BetterNameTextureMTL& operator=(const BetterNameTextureMTL&) = delete;
+};
+
+/// @brief A class that holds the mutable state of the DrawableTextureMTL.
+class DrawableHolder {
+ public:
+  DrawableHolder() = default;
+
+  ~DrawableHolder() = default;
+
+  bool AcquireNextDrawable(CAMetalLayer* layer) {
```

Comment:
This seems to be a couple of levels of overkill. All we need is a lambda that is lazily evaluated and its results cached. `std::async` with `std::launch::deferred` will defer calling the lambda on the calling thread when the future is awaited upon. And, the future itself will hold the value around till you discard it.

So the `DrawableHolder` can just be `std::future<std::shared_ptr<impeller::Texture>>` (typedef it maybe). That can be obtained using a call to `std::async(std::launch::deferred, [layer](){return WrapIntoAtexture([layer nextDrawable]);})`

For style points, `TextureMTL` itself can take only a single argument which is a lambda that returns a texture. A lambda that will always be evaluated lazily. And in case you already have a texture, the lambda becomes `[texture](){return texture};`. Create a factory that creates the lambda if need be to keep the interface the same.

That way, you don't have to create a subclass hierarchy to LayerMTL. More importantly, you also don't have to think about naming the subclasses.

Reply:
Unfortunately I need access to both the texture and its drawable, the former to add to the MTL attachment, the later to present in SurfaceMTL.

Reply:
I can definitely compress this into a subclass, but I can't think of a way to do so without adding new API to query the drawable state.

Reply:
> Unfortunately I need access to both the texture and its drawable, the former to add to the MTL attachment, the later to present in SurfaceMTL.

Thats just a future to a `std::pair` or `struct`.

Reply:
But the usage of the drawable isn't internal to the texture or to the callback, its done by the surface MTL. So roughly:

1. surface_mtl has access to CAMetalLayer.
2. Create TextureMTL from CAMetalLayer.
3a. Worker thread begins encoding root pass and waits on next drawable.
3b. surface_mtl waits on same drawable (without requesting it twice) and presents.

But thinking about it, I could probably have an object captured by that callback manage this state (essentially the state holder object I created)?

So I'll give that a shot. We do stil need an additional bit for whether or not the texture is lazy